builtin/plugins/chainconfig/chainconfig.go

```diff
@@ -542,3 +558,70 @@ func removeFeature(ctx contract.Context, name string) error {
 }
 
 var Contract plugin.Contract = contract.MakePluginContract(&ChainConfig{})
+
+func (c *ChainConfig) SetValidatorInfo(ctx contract.Context, req *SetValidatorInfoRequest) error {
+	if !ctx.FeatureEnabled(loomchain.ChainCfgVersion1_2, false) {
+		return ErrFeatureNotEnabled
+	}
+
+	if req.BuildNumber == 0 {
+		return ErrInvalidRequest
+	}
+	senderAddr := ctx.Message().Sender
+	validators, err := getCurrentValidators(ctx)
+	if err != nil {
+		return err
+	}
+	isValidator := false
+	for _, validator := range validators {
+		if validator.Compare(senderAddr) == 0 {
+			isValidator = true
+			break
+		}
+	}
+	if !isValidator {
+		return ErrNotAuthorized
+	}
+
+	validator := &ValidatorInfo{
+		Address:     senderAddr.MarshalPB(),
+		BuildNumber: req.BuildNumber,
+		UpdatedAt:   uint64(time.Now().Unix()),
```

Comment:
Never use `time.Now()` in a contract, it will differ between nodes, thus changing persistent state in a slightly different way on every node in the cluster, since the state change is inconsistent the nodes will never reach consensus and the chain will halt. Though it's strange the e2e test hasn't failed.

You want to use `ctx.Now()` instead, that's the block timestamp, which is going to be consistent across the cluster.